File: apps/api/app/schemas/real/relationship_intelligence.py
"""Relationship intelligence schemas (real application layer, Phase 21).

Mirror the shape the frontend ``@trinetra/types`` consumes for relationship
intelligence while keeping UUID identities and additive semantics. These are
read/compute views: nothing here accepts actor identity from the client.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field

from app.schemas.common import SchemaBase


class RelationshipObservationRead(SchemaBase):
    id: UUID
    relationship_id: UUID
    source_dataset: str
    source_type: str | None = None
    source_record: str | None = None
    extraction_method: str | None = None
    observed_at: datetime | None = None
    confidence: float | None = None
    evidence_refs: list[str] = Field(default_factory=list)
    provenance: bool = True


class RelationshipConflictRead(SchemaBase):
    type: str
    field: str
    sources: list[str] = Field(default_factory=list)
    values: list[Any] = Field(default_factory=list)
    detail: str | None = None


class RelationshipEvidenceSupportRead(SchemaBase):
    relationship_id: UUID
    evidence_count: int = 0
    evidence_ids: list[UUID] = Field(default_factory=list)
    datasets: list[str] = Field(default_factory=list)
    linked: bool = False
    message: str = "No linked evidence"


class RelationshipIntelligenceRead(SchemaBase):
    relationship_id: UUID
    investigation_id: UUID
    source_entity_id: UUID
    source_entity_name: str | None = None
    source_entity_type: str | None = None
    target_entity_id: UUID
    target_entity_name: str | None = None
    target_entity_type: str | None = None
    relationship_type: str
    direction: str = "undirected"
    confidence: float = 0.0
    source: str | None = None
    evidence_refs: list[str] = Field(default_factory=list)
    verification_status: str | None = None
    extraction_method: str | None = None
    intelligence_status: str = "observed"
    linkage_score: float = 0.0
    confidence_label: str = "LOW"
    correlation_version: str = "relationship-intelligence-v1"
    correlation_key: str | None = None
    observation_count: int = 0
    source_count: int = 0
    observation_predicate: str = "observation"
    first_observed_at: datetime | None = None
    last_observed_at: datetime | None = None
    conflict_flags: list[dict[str, Any]] = Field(default_factory=list)
    observations: list[RelationshipObservationRead] = Field(default_factory=list)
    evidence: RelationshipEvidenceSupportRead
    verified_by: str | None = None
    verified_at: datetime | None = None
    rejection_reason: str | None = None
    metadata_: dict[str, Any] = Field(default_factory=dict, serialization_alias="metadata")
    created_at: datetime
    updated_at: datetime


class RelationshipIntelligenceItemRead(SchemaBase):
    """Lightweight list item (no full observations/evidence payloads)."""

    relationship_id: UUID
    investigation_id: UUID
    source_entity_id: UUID
    source_entity_name: str | None = None
    source_entity_type: str | None = None
    target_entity_id: UUID
    target_entity_name: str | None = None
    target_entity_type: str | None = None
    relationship_type: str
    direction: str = "undirected"
    confidence: float = 0.0
    verification_status: str | None = None
    extraction_method: str | None = None
    intelligence_status: str = "observed"
    linkage_score: float = 0.0
    confidence_label: str = "LOW"
    correlation_version: str = "relationship-intelligence-v1"
    correlation_key: str | None = None
    observation_count: int = 0
    source_count: int = 0
    observation_predicate: str = "observation"
    first_observed_at: datetime | None = None
    last_observed_at: datetime | None = None
    conflict_count: int = 0
    evidence_count: int = 0
    verified_by: str | None = None
    verified_at: datetime | None = None
    rejection_reason: str | None = None


class RelationshipIntelligenceListRead(SchemaBase):
    investigation_id: UUID
    items: list[RelationshipIntelligenceItemRead] = Field(default_factory=list)
    algorithm_version: str
    evaluated_at: datetime


class RelationshipObservationsRead(SchemaBase):
    relationship_id: UUID
    investigation_id: UUID
    observations: list[RelationshipObservationRead] = Field(default_factory=list)
    algorithm_version: str
    evaluated_at: datetime


class RelationshipConfirmRequest(SchemaBase):
    reason: str | None = None


class RelationshipRejectRequest(SchemaBase):
    reason: str | None = None


class RelationshipEvaluationResponse(SchemaBase):
    investigation_id: UUID
    evaluated_relationships: int = 0
    correlation_groups: int = 0
    correlated_relationships: int = 0
    conflicts_detected: int = 0
    algorithm_version: str
    evaluated_at: datetime

File: apps/web/src/lib/api/relationship-intelligence.ts
/**
 * Typed client functions for the relationship-intelligence API (/api/v2).
 *
 * These mirror the backend routes added during Phase 21:
 *   GET    /relationships/{id}/intelligence
 *   GET    /relationships/{id}/observations
 *   GET    /relationships/{id}/evidence
 *   POST   /relationships/{id}/confirm
 *   POST   /relationships/{id}/reject
 *   POST   /investigations/{id}/relationships/evaluate
 *   GET    /investigations/{id}/relationships/intelligence
 *
 * Response shapes follow the backend wire format (snake_case). Mapping to the
 * frontend camelCase intelligence types happens in relationship-intelligence.service.
 */

import { API_BASE_URL } from './config';
import { apiFetch } from './client';
import type { JsonObject, Uuid } from './investigations';

export type RealIntelligenceStatus = 'NEEDS_REVIEW' | 'REVIEWED' | 'DISCARDED';
export type RealConfidenceLabel = 'HIGH' | 'MEDIUM' | 'LOW';

export interface RealRelationshipObservation {
  observation_id: string;
  source_id: string;
  source_label: string;
  reference?: string | null;
  observed_at?: string | null;
  confidence: number;
}

export interface RealRelationshipEvidenceSupportRead {
  relationship_id: Uuid;
  investigation_id: Uuid;
  evidence_count: number;
  evidence_ids: Uuid[];
  datasets: string[];
  linked: boolean;
  message: string;
}

export interface RealRelationshipIntelligence {
  relationship_id: Uuid;
  investigation_id: Uuid;
  source_entity_id: Uuid;
  source_entity_name: string | null;
  source_entity_type: string | null;
  target_entity_id: Uuid;
  target_entity_name: string | null;
  target_entity_type: string | null;
  relationship_type: string;
  direction: 'directed' | 'undirected';
  confidence: number;
  source: string | null;
  evidence_refs: string[];
  verification_status: string | null;
  extraction_method: string | null;
  intelligence_status: RealIntelligenceStatus;
  linkage_score: number;
  confidence_label: RealConfidenceLabel;
  correlation_version: string;
  correlation_key: string;
  observation_count: number;
  source_count: number;
  observation_predicate: string | null;
  first_observed_at: string | null;
  last_observed_at: string | null;
  conflict_flags: string[];
  observations: RealRelationshipObservation[];
  evidence: RealRelationshipEvidenceSupportRead;
  verified_by: Uuid | null;
  verified_at: string | null;
  rejection_reason: string | null;
  metadata_: JsonObject;
  created_at: string;
  updated_at: string;
}

export interface RealRelationshipIntelligenceItem {
  relationship_id: Uuid;
  investigation_id: Uuid;
  source_entity_id: Uuid;
  source_entity_name: string | null;
  source_entity_type: string | null;
  target_entity_id: Uuid;
  target_entity_name: string | null;
  target_entity_type: string | null;
  relationship_type: string;
  direction: 'directed' | 'undirected';
  confidence: number;
  verification_status: string | null;
  extraction_method: string | null;
  intelligence_status: RealIntelligenceStatus;
  linkage_score: number;
  confidence_label: RealConfidenceLabel;
  correlation_version: string;
  correlation_key: string;
  observation_count: number;
  source_count: number;
  observation_predicate: string | null;
  first_observed_at: string | null;
  last_observed_at: string | null;
  conflict_count: number;
  evidence_count: number;
  verified_by: Uuid | null;
  verified_at: string | null;
  rejection_reason: string | null;
}

export interface RealRelationshipIntelligenceList {
  investigation_id: Uuid;
  items: RealRelationshipIntelligenceItem[];
  algorithm_version: string;
  evaluated_at: string;
}

export interface RealRelationshipEvaluationResponse {
  investigation_id: Uuid;
  evaluated_relationships: number;
  correlation_groups: number;
  correlated_relationships: number;
  conflicts_detected: number;
  algorithm_version: string;
  evaluated_at: string;
}

export interface RealRelationshipObservationsResponse {
  relationship_id: Uuid;
  investigation_id: Uuid;
  observations: RealRelationshipObservation[];
  algorithm_version: string;
  evaluated_at: string;
}

export async function getRelationshipIntelligence(
  relationshipId: Uuid,
): Promise<RealRelationshipIntelligence> {
  return apiFetch<RealRelationshipIntelligence>(
    API_BASE_URL,
    `/relationships/${relationshipId}/intelligence`,
  );
}

export async function getRelationshipObservations(
  relationshipId: Uuid,
): Promise<RealRelationshipObservationsResponse> {
  return apiFetch<RealRelationshipObservationsResponse>(
    API_BASE_URL,
    `/relationships/${relationshipId}/observations`,
  );
}

export async function getRelationshipEvidence(
  relationshipId: Uuid,
): Promise<RealRelationshipEvidenceSupportRead> {
  return apiFetch<RealRelationshipEvidenceSupportRead>(
    API_BASE_URL,
    `/relationships/${relationshipId}/evidence`,
  );
}

export async function listRelationshipIntelligence(
  investigationId: Uuid,
): Promise<RealRelationshipIntelligenceList> {
  return apiFetch<RealRelationshipIntelligenceList>(
    API_BASE_URL,
    `/investigations/${investigationId}/relationships/intelligence`,
  );
}

export async function evaluateRelationshipIntelligence(
  investigationId: Uuid,
): Promise<RealRelationshipEvaluationResponse> {
  return apiFetch<RealRelationshipEvaluationResponse>(
    API_BASE_URL,
    `/investigations/${investigationId}/relationships/evaluate`,
    { method: 'POST' },
  );
}

export async function confirmRelationship(
  relationshipId: Uuid,
  reason?: string,
): Promise<RealRelationshipIntelligence> {
  return apiFetch<RealRelationshipIntelligence>(
    API_BASE_URL,
    `/relationships/${relationshipId}/confirm`,
    { method: 'POST', body: { reason: reason ?? null } },
  );
}

export async function rejectRelationship(
  relationshipId: Uuid,
  reason?: string,
): Promise<RealRelationshipIntelligence> {
  return apiFetch<RealRelationshipIntelligence>(
    API_BASE_URL,
    `/relationships/${relationshipId}/reject`,
    { method: 'POST', body: { reason: reason ?? null } },
  );
}
